modules/data.atmosphere/R/download.MACA.R

```diff
@@ -0,0 +1,145 @@
+##' @title download.MACA
+##' @export
+##' @param outfolder
+##' @param start_date , of the format "YEAR-01-01 00:00:00"
+##' @param end_date , of the format "YEAR-12-31 23:59:59"
+##' @param lat
+##' @param lon
+##' @param model , select which MACA model to run (options are BNU-ESM, CNRM-CM5, CSIRO-Mk3-6-0, bcc-csm1-1, bcc-csm1-1-m, CanESM2, GFDL-ESM2M, GFDL-ESM2G, HadGEM2-CC365, HadGEM2-ES365, inmcm4, MIROC5, MIROC-ESM, MIROC-ESM-CHEM, MRI-CGCM3, CCSM4, IPSL-CM5A-LR, IPSL-CM5A-MR, IPSL-CM5B-LR, NorESM1-M)
+##' @param scenario , select which scenario to run (options are rcp45, rcp85)
+##' @param ensemble_member , r1i1p1 is the only ensemble member available for this dataset, CCSM4 uses r6i1p1 instead
+##'
+##' @author James Simkins
+download.MACA <- function(outfolder, start_date, end_date, site_id, lat.in, lon.in, overwrite=FALSE, verbose=FALSE, model='IPSL-CM5A-LR', scenario='rcp85', ensemble_member='r1i1p1', ...){  
+  require(PEcAn.utils)
+  require(lubridate)
+  require(ncdf4)
+  start_date <- as.POSIXlt(start_date, tz = "GMT")
+  end_date <- as.POSIXlt(end_date, tz = "GMT")
+  start_year <- year(start_date)
+  end_year   <- year(end_date)
+  site_id = as.numeric(site_id)
+  model = paste0(model)
+  scenario = paste0(scenario)
+  ensemble_member = paste0(ensemble_member)
+  outfolder = paste0(outfolder,"_site_",paste0(site_id %/% 1000000000, "-", site_id %% 1000000000))
+  
+  if (model == 'CCSM4'){
+    ensemble_member = 'r6i1p1'
+  }
+  
+  lat.in = as.numeric(lat.in)
+  lat = lat.in - 25.063077926635742
+  lat_MACA = round(lat*24)
+  lon.in = as.numeric(lon.in)
+  if (lon.in < 0){
+    lon.in = 180 - lon.in}
+  lon = lon.in - 235.22784423828125
+  lon_MACA = round(lon*24)
+  
+
+  dap_base ='http://thredds.northwestknowledge.net:8080/thredds/dodsC/MACAV2'
+
+  ylist <- seq(start_year,end_year,by=1)
+  rows = length(ylist)
+  results <- data.frame(file=character(rows), host=character(rows),
+                        mimetype=character(rows), formatname=character(rows),
+                        startdate=character(rows), enddate=character(rows),
+                        dbfile.name = paste("MACA",model,scenario,ensemble_member,sep="."),#"MACA",
+                        stringsAsFactors = FALSE)
+  
+  
+  prevar = data.frame(DAP.name = c("tasmax","tasmin","rsds","uas","vas","huss","pr"), 
+                      CF.name = c("air_temperature","air_temperature","surface_downwelling_shortwave_flux_in_air","eastward_wind","northward_wind","specific_humidity","precipitation"),
+                      units = c('Kelvin','Kelvin',"W/m2","m/s","m/s","g/g","kg/m2/s")
+                      )
+  var = data.frame(DAP.name = c("air_temperature", "rsds","uas","vas","huss","pr"),
+                   CF.name = c("air_temperature","surface_downwelling_shortwave_flux_in_air","eastward_wind","northward_wind","specific_humidity","precipitation_flux"),
+                   units = c('Kelvin',"W/m2","m/s","m/s","g/g","kg/m2/s")
+  )
+  
+  for (i in 1:rows){
```

Comment:
It looks like you're looping over years and then downloading a 5 year block within the loop. That means you'll typically download the same 5 year block 5 times. Seems like it would be better to loop over blocks as the outer loop, do the download, and then loop over years internally.